Feature/wp error notifications

ulferts opened the pull request:
Employs the NotificationService for all Angular based messages to the user. 

By using the NotificationService, we can remove the FlashMessageDirective which was used to mimic the rails based flash messages.

The PR does not aim to improve the error handling. There are a lot of places where the error handling is still lacking, but they are not part of this PR. 

Fixes: 
https://community.openproject.org/work_packages/20922
Addresses part of: 
https://community.openproject.org/work_packages/20248 

Todo:
- [x] remove last remains of `scope.emit('flashMessage', ...`

ulferts committed:
revamp flash directive to transmit to NotificationsHelper

This is only a temporary reworking as it is now wrong to use a directive

ulferts committed:
replace flash-message directive by NotificationsService

The NotificationsService renders the flash-message directive irrelevant and it additionally displays the notification much nicer than the directive.

Additionally, an ApiNotificationsService is introduced that is able to cover error formats provided by the experimental api-

ulferts committed:
remove flash-message from angular layout

ulferts committed:
adapt karama tests

ulferts committed:
separate NotificationsService and ApiHelper

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://ulferts-feature-wp-error-notifications.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://ulferts-feature-wp-error-notifications.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts committed:
adapt spec, introduce features/page_objects

The features/page_objects folder should house all the page objects used within specs

ulferts committed:
specs for api-notifications-service

ulferts committed:
replaces ApiHelper.handleError by ApiNotificationsService

ulferts committed:
remove flash-message usage from time_entries

ulferts committed:
code styling

ulferts committed:
misc styling fixes

ulferts committed:
cleanup injections

The injections referenced are no longer in use.

NobodysNightmare commented on frontend/app/services/api-notifications-service.js:
```diff
@@ -26,37 +26,26 @@
+      NotificationsService.addError('', messages);
+    }
+    else {
+      NotificationsService.addError(messages[0], []);
```
This if-else looks a little clumsy.
I think to remember how it was intended: "we have a list of error messages and maybe even an additional message", but this is not how you are using it at the moment.

Two (and a half) ideas:
- always use the errors list
  - either always passing a message (`addError('The following fancy errors occured', messages)`)
  - only passing the errors (second param) and deciding inside the `NotificationService` how to display the error message depending on error count and presence of a message
- consequently change the signature (and assumptions) of the `NotificationService` to support how we are now using it (e.g. it explicitly checks whether errors were passed, because it did **not** intend to have no errors)

My favorite option at the moment is option 1.2, since it seems most consistent to me...

ulferts replied:
If we move the decision of how to display the error message into the NotificationService we are in effect moving the `if` statement in there. Now that might be an improvement if we see a lot of places that call `addError` having an `if` statement before. 

I am OK with removing the necessity of having `errors`, the `message` should suffice (maybe the wording is bad as well). That would remove the awkward empty array on the `else` call.

But I think that we should otherwise leave the interface as it is. Because then we can decide on which places we want to invest into good error messages, e.g. errors on wp create/update and on which places we roll with what the API gives us, at least for the time being.

So if it is fine with you, I'd remove the necessity of having `errors`. Other than that, I wouldn't want to change it now.

ulferts replied:
If we move the decision of how to display the error message into the NotificationService we are in effect moving the `if` statement in there. Now that might be an improvement if we see a lot of places that call `addError` having an `if` statement before. 

I am OK with removing the necessity of having `errors`, the `message` should suffice (maybe the wording is bad as well). That would remove the awkward empty array on the `else` call.

But I think that we should otherwise leave the interface as it is. Because then we can decide on which places we want to invest into good error messages, e.g. errors on wp create/update and on which places we roll with what the API gives us, at least for the time being.

So if it is fine with you, I'd remove the necessity of having `errors`. Other than that, I wouldn't want to change it now.

NobodysNightmare replied:
@ulferts This sounds reasonable. My main concern was really the repetitive use of the empty second parameter. :+1:

NobodysNightmare commented on frontend/app/work_packages/helpers/api-helper.js:
```diff
@@ -51,14 +43,16 @@ module.exports = function() {
     },
 
+    getErrorMessage: function(error) {
+      return ApiHelper.getErrorMessages(error).join(' ');
```
First of all: Thank you for introducing `getErrorMessages` :+1: 

But secondly: Who would ever need error messages joined by a single space character? This sounds like the least displayable error string imaginable (except for the same string reversed, maybe).

Okay, I digged a little. The only caller seems to be [here](https://github.com/opf/openproject/blob/4889b8615c51e877b19a5a066f04cdebf494819b/frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js#L160). Then it is used downstream [here](https://github.com/opf/openproject/blob/4889b8615c51e877b19a5a066f04cdebf494819b/frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js#L45).

The only reason that this did not produce ugly errors yet, is that when there is a general (`_common`) error, the API will AFAIK only ever return a single error. And apparently there are no (observed) cases of more than one error per field, because that would also blow up.

NobodysNightmare commented:
Not sure if this is caused by your PR, but existing work packages with validation error behave completely strange:

## Reproduction
1. set the subject to be empty
2. do a good change
3. press the save button on the subject field
4. press the save button on the "good change"

## Expected
- for both buttons the same validation errors are displayed
- for both buttons the WP is not saved

### Actual
- for the subject button, the WP is not saved and an error is shown
- for the "good" button, the WP **is** saved (with the old subject) and no error is shown

If I see it correctly this will also work for other validation errors.

If you don't think that this bug belongs to your PR, I can turn the above report into a separate work package, but I think this is a critical bug for the release (IMO).

edit: okay, this one is also present on the dev...

NobodysNightmare commented:
Okay... I am tripping over a lot of bugs in this review, but according to @RobinWagner all of my errors found so far are already present on `dev` and reported :/

basiszwo commented:
Do we want to merge it as is or do we need further development?

NobodysNightmare commented:
My code comments still remain. I think the intention of `addError` and the actual usages (introduced with this PR) differ quite much. This should be harmonized, one way or another.

As far as I can tell, this PR does functionally what it should.

ulferts committed:
remove helper method no longer needed

ulferts committed:
allows for creating error notifications with only an error message

ulferts commented:
![image](https://cloud.githubusercontent.com/assets/617519/9494047/14c18446-4c04-11e5-8b87-5670589d50d9.png)

if travis agrees.

NobodysNightmare commented:
Now it LGTM :+1: 

Waiting for travis nevertheless.
